File: app/modelos/dominios.py
"""Catalogos versionados: tipos, percentuais, fundamentacoes, agentes, textos."""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    ForeignKey,
    Integer,
    Numeric,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.modelos.base import Base, DataPura


class TipoAdicional(Base):
    __tablename__ = "tipo_adicional"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(24), nullable=False, unique=True)
    nome: Mapped[str] = mapped_column(String(80), nullable=False)
    # como entra na frase da recomendacao: 'adicional de insalubridade'
    nome_recomendacao: Mapped[str] = mapped_column(String(80), nullable=False)
    base_legal: Mapped[str | None] = mapped_column(Text)
    ativo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    percentuais: Mapped[list["PercentualAplicavel"]] = relationship(
        back_populates="tipo_adicional"
    )


class TipoMovimento(Base):
    __tablename__ = "tipo_movimento"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(20), nullable=False, unique=True)
    nome: Mapped[str] = mapped_column(String(40), nullable=False)
    gera_direito: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    ativo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


class TipoRisco(Base):
    __tablename__ = "tipo_risco"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(20), nullable=False, unique=True)
    nome: Mapped[str] = mapped_column(String(60), nullable=False)


class PercentualAplicavel(Base):
    __tablename__ = "percentual_aplicavel"

    id: Mapped[int] = mapped_column(primary_key=True)
    tipo_adicional_id: Mapped[int] = mapped_column(
        ForeignKey("tipo_adicional.id"), nullable=False
    )
    grau: Mapped[str] = mapped_column(String(10), nullable=False)
    rotulo: Mapped[str] = mapped_column(String(40), nullable=False)
    valor: Mapped[Decimal] = mapped_column(Numeric(5, 2), nullable=False)
    base_calculo: Mapped[str] = mapped_column(
        String(60), nullable=False, default="vencimento do cargo efetivo"
    )

    tipo_adicional: Mapped[TipoAdicional] = relationship(
        back_populates="percentuais", lazy="selectin"
    )

    __table_args__ = (
        UniqueConstraint("tipo_adicional_id", "grau", name="uq_percentual"),
        CheckConstraint(
            "grau IN ('MINIMO','MEDIO','MAXIMO','UNICO')", name="ck_grau"
        ),
    )


class FundamentacaoLegal(Base):
    __tablename__ = "fundamentacao_legal"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(30), nullable=False, unique=True)
    norma: Mapped[str] = mapped_column(String(40), nullable=False)
    anexo: Mapped[str | None] = mapped_column(String(20))
    tipo_risco_id: Mapped[int | None] = mapped_column(ForeignKey("tipo_risco.id"))
    texto: Mapped[str] = mapped_column(Text, nullable=False)
    dispositivo_conferido_em: Mapped[date | None] = mapped_column(DataPura)
    vigente: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    tipo_risco: Mapped[TipoRisco | None] = relationship(lazy="selectin")


class AgenteNocivo(Base):
    __tablename__ = "agente_nocivo"

    id: Mapped[int] = mapped_column(primary_key=True)
    descricao: Mapped[str] = mapped_column(String(200), nullable=False, unique=True)
    tipo_risco_id: Mapped[int] = mapped_column(ForeignKey("tipo_risco.id"), nullable=False)
    fundamentacao_id: Mapped[int | None] = mapped_column(
        ForeignKey("fundamentacao_legal.id")
    )
    percentual_sugerido_id: Mapped[int | None] = mapped_column(
        ForeignKey("percentual_aplicavel.id")
    )
    exige_reavaliacao_quantitativa: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    # sinonimo -> canonico (de-para que altera semantica: exige aprovacao)
    agente_canonico_id: Mapped[int | None] = mapped_column(ForeignKey("agente_nocivo.id"))
    ativo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    tipo_risco: Mapped[TipoRisco] = relationship(lazy="selectin")
    fundamentacao: Mapped[FundamentacaoLegal | None] = relationship(lazy="selectin")


class TextoPadrao(Base):
    __tablename__ = "texto_padrao"

    id: Mapped[int] = mapped_column(primary_key=True)
    categoria: Mapped[str] = mapped_column(String(30), nullable=False)
    codigo: Mapped[str] = mapped_column(String(40), nullable=False)
    template: Mapped[str] = mapped_column(Text, nullable=False)
    versao: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    vigente: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    # false = paragrafo estatico do .docx; existe so como referencia de catalogo
    renderizado_pelo_modelo: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True
    )
    dispositivo_conferido_em: Mapped[date | None] = mapped_column(DataPura)

    __table_args__ = (
        UniqueConstraint("categoria", "codigo", "versao", name="uq_texto"),
        CheckConstraint(
            "categoria IN ('ALTERACAO','RECOMENDACAO','REAVALIACAO',"
            "'RODAPE_RESPONSABILIDADE','ASSUNTO','PREAMBULO','ENCERRAMENTO')",
            name="ck_cat",
        ),
    )


class TipoMarcoInicial(Base):
    __tablename__ = "tipo_marco_inicial"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(30), nullable=False, unique=True)
    rotulo: Mapped[str] = mapped_column(String(80), nullable=False)
    base_legal: Mapped[str | None] = mapped_column(Text)


class TipoProcesso(Base):
    __tablename__ = "tipo_processo"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(30), nullable=False, unique=True)
    nome: Mapped[str] = mapped_column(String(80), nullable=False)
    cor_hex: Mapped[str | None] = mapped_column(String(7))
    modulo: Mapped[str] = mapped_column(String(32), nullable=False, default="ADICIONAL")


class Parametro(Base):
    """Parametros operacionais editaveis em /config (SLA, por exemplo).

    Fica no banco, nao no .env: o .env e infraestrutura (porta, caminho, chave);
    isto e regra de trabalho do setor, que o coordenador muda sozinho.
    """

    __tablename__ = "parametro"

    chave: Mapped[str] = mapped_column(String(60), primary_key=True)
    valor: Mapped[str] = mapped_column(Text, nullable=False)
    descricao: Mapped[str | None] = mapped_column(Text)


class ChecklistModelo(Base):
    """Modelo de checklist aplicavel a um processo (catalogo /checklists-modelo)."""

    __tablename__ = "checklist_modelo"

    id: Mapped[int] = mapped_column(primary_key=True)
    nome: Mapped[str] = mapped_column(String(120), nullable=False, unique=True)
    tipo_processo_id: Mapped[int | None] = mapped_column(ForeignKey("tipo_processo.id"))
    estado_alvo: Mapped[str | None] = mapped_column(String(32))
    # um item por linha, na ordem
    itens: Mapped[str] = mapped_column(Text, nullable=False)
    ativo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    tipo_processo: Mapped["TipoProcesso | None"] = relationship(lazy="selectin")

    @property
    def lista_de_itens(self) -> list[str]:
        return [linha.strip() for linha in (self.itens or "").splitlines() if linha.strip()]


class FluxoEtapa(Base):
    __tablename__ = "fluxo_etapa"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(30), nullable=False, unique=True)
    nome: Mapped[str] = mapped_column(String(60), nullable=False)
    ordem: Mapped[int] = mapped_column(SmallInteger, nullable=False, unique=True)
    tipo: Mapped[str] = mapped_column(String(12), nullable=False, default="FLUXO")
    terminal: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    __table_args__ = (
        CheckConstraint("tipo IN ('FLUXO','REPOSITORIO')", name="ck_etapa_tipo"),
    )


__all__ = [
    "AgenteNocivo",
    "ChecklistModelo",
    "FluxoEtapa",
    "Parametro",
    "FundamentacaoLegal",
    "PercentualAplicavel",
    "TextoPadrao",
    "TipoAdicional",
    "TipoMarcoInicial",
    "TipoMovimento",
    "TipoProcesso",
    "TipoRisco",
]

File: nuvem/src/db/esquema/dominios.ts
/**
 * Catalogos versionados: tipos, percentuais, fundamentacoes, agentes, textos.
 * Porte de `app/modelos/dominios.py`.
 *
 * Convencao de todo o esquema: nome da tabela, nome da coluna E nome da
 * propriedade TypeScript identicos ao Python, em snake_case. Servicos e
 * templates estao sendo portados mecanicamente, e `processo.numero_sei` tem de
 * continuar sendo `processo.numero_sei` do outro lado.
 *
 * `String(n)` vira `varchar(n)`: o DDL de referencia do projeto sempre foi
 * PostgreSQL, e no SQLite o limite era so decorativo. Aqui ele passa a valer.
 */

import { sql } from 'drizzle-orm';
import {
  type AnyPgColumn,
  boolean,
  check,
  integer,
  numeric,
  pgTable,
  smallint,
  text,
  unique,
  varchar,
} from 'drizzle-orm/pg-core';

import { DataPura } from './base';

export const tipo_adicional = pgTable('tipo_adicional', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 24 }).notNull().unique(),
  nome: varchar('nome', { length: 80 }).notNull(),
  // como entra na frase da recomendacao: 'adicional de insalubridade'
  nome_recomendacao: varchar('nome_recomendacao', { length: 80 }).notNull(),
  base_legal: text('base_legal'),
  ativo: boolean('ativo').notNull().default(true),
});

export const tipo_movimento = pgTable('tipo_movimento', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 20 }).notNull().unique(),
  nome: varchar('nome', { length: 40 }).notNull(),
  gera_direito: boolean('gera_direito').notNull().default(true),
  ativo: boolean('ativo').notNull().default(true),
});

export const tipo_risco = pgTable('tipo_risco', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 20 }).notNull().unique(),
  nome: varchar('nome', { length: 60 }).notNull(),
});

export const percentual_aplicavel = pgTable(
  'percentual_aplicavel',
  {
    id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
    tipo_adicional_id: integer('tipo_adicional_id')
      .notNull()
      .references(() => tipo_adicional.id),
    grau: varchar('grau', { length: 10 }).notNull(),
    rotulo: varchar('rotulo', { length: 40 }).notNull(),
    valor: numeric('valor', { precision: 5, scale: 2, mode: 'string' }).notNull(),
    base_calculo: varchar('base_calculo', { length: 60 })
      .notNull()
      .default('vencimento do cargo efetivo'),
  },
  (t) => [
    unique('uq_percentual').on(t.tipo_adicional_id, t.grau),
    check('ck_grau', sql.raw(`grau IN ('MINIMO','MEDIO','MAXIMO','UNICO')`)),
  ],
);

export const fundamentacao_legal = pgTable('fundamentacao_legal', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 30 }).notNull().unique(),
  norma: varchar('norma', { length: 40 }).notNull(),
  anexo: varchar('anexo', { length: 20 }),
  tipo_risco_id: integer('tipo_risco_id').references(() => tipo_risco.id),
  texto: text('texto').notNull(),
  dispositivo_conferido_em: DataPura('dispositivo_conferido_em'),
  vigente: boolean('vigente').notNull().default(true),
});

export const agente_nocivo = pgTable('agente_nocivo', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  descricao: varchar('descricao', { length: 200 }).notNull().unique(),
  tipo_risco_id: integer('tipo_risco_id')
    .notNull()
    .references(() => tipo_risco.id),
  fundamentacao_id: integer('fundamentacao_id').references(() => fundamentacao_legal.id),
  percentual_sugerido_id: integer('percentual_sugerido_id').references(
    () => percentual_aplicavel.id,
  ),
  exige_reavaliacao_quantitativa: boolean('exige_reavaliacao_quantitativa')
    .notNull()
    .default(false),
  // sinonimo -> canonico (de-para que altera semantica: exige aprovacao)
  agente_canonico_id: integer('agente_canonico_id').references((): AnyPgColumn => agente_nocivo.id),
  ativo: boolean('ativo').notNull().default(true),
});

export const texto_padrao = pgTable(
  'texto_padrao',
  {
    id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
    categoria: varchar('categoria', { length: 30 }).notNull(),
    codigo: varchar('codigo', { length: 40 }).notNull(),
    template: text('template').notNull(),
    versao: integer('versao').notNull().default(1),
    vigente: boolean('vigente').notNull().default(true),
    // false = paragrafo estatico do .docx; existe so como referencia de catalogo
    renderizado_pelo_modelo: boolean('renderizado_pelo_modelo').notNull().default(true),
    dispositivo_conferido_em: DataPura('dispositivo_conferido_em'),
  },
  (t) => [
    unique('uq_texto').on(t.categoria, t.codigo, t.versao),
    check(
      'ck_cat',
      sql.raw(
        `categoria IN ('ALTERACAO','RECOMENDACAO','REAVALIACAO',` +
          `'RODAPE_RESPONSABILIDADE','ASSUNTO','PREAMBULO','ENCERRAMENTO')`,
      ),
    ),
  ],
);

export const tipo_marco_inicial = pgTable('tipo_marco_inicial', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 30 }).notNull().unique(),
  rotulo: varchar('rotulo', { length: 80 }).notNull(),
  base_legal: text('base_legal'),
});

export const tipo_processo = pgTable('tipo_processo', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  codigo: varchar('codigo', { length: 30 }).notNull().unique(),
  nome: varchar('nome', { length: 80 }).notNull(),
  cor_hex: varchar('cor_hex', { length: 7 }),
  modulo: varchar('modulo', { length: 32 }).notNull().default('ADICIONAL'),
});

/**
 * Parametros operacionais editaveis em /config (SLA, por exemplo).
 *
 * Fica no banco, nao em variavel de ambiente: ambiente e infraestrutura (porta,
 * URL, chave); isto e regra de trabalho do setor, que o coordenador muda
 * sozinho.
 */
export const parametro = pgTable('parametro', {
  chave: varchar('chave', { length: 60 }).primaryKey(),
  valor: text('valor').notNull(),
  descricao: text('descricao'),
});

/** Modelo de checklist aplicavel a um processo (catalogo /checklists-modelo). */
export const checklist_modelo = pgTable('checklist_modelo', {
  id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
  nome: varchar('nome', { length: 120 }).notNull().unique(),
  tipo_processo_id: integer('tipo_processo_id').references(() => tipo_processo.id),
  estado_alvo: varchar('estado_alvo', { length: 32 }),
  // um item por linha, na ordem (ver `lista_de_itens` em src/dominio/dominios.ts)
  itens: text('itens').notNull(),
  ativo: boolean('ativo').notNull().default(true),
});

export const fluxo_etapa = pgTable(
  'fluxo_etapa',
  {
    id: integer('id').primaryKey().generatedByDefaultAsIdentity(),
    codigo: varchar('codigo', { length: 30 }).notNull().unique(),
    nome: varchar('nome', { length: 60 }).notNull(),
    ordem: smallint('ordem').notNull().unique(),
    tipo: varchar('tipo', { length: 12 }).notNull().default('FLUXO'),
    terminal: boolean('terminal').notNull().default(false),
  },
  () => [check('ck_etapa_tipo', sql.raw(`tipo IN ('FLUXO','REPOSITORIO')`))],
);
